src/variable.cpp

```diff
@@ -1248,15 +1249,31 @@ Variable concatenate(const Variable &a1, const Variable &a2, const Dim dim) {
 
 Variable rebin(const Variable &var, const Variable &oldCoord,
                const Variable &newCoord) {
-  auto rebinned(var);
-  auto dims = rebinned.dimensions();
+  dataset::expect::countsOrCountsDensity(var);
   const Dim dim = coordDimension[newCoord.tag().value()];
-  dims.resize(dim, newCoord.dimensions()[dim] - 1);
-  rebinned.setDimensions(dims);
-  // TODO take into account unit if values have been divided by bin width.
-  require<FloatingPointVariableConcept>(rebinned.data())
-      .rebin(var.data(), dim, oldCoord.data(), newCoord.data());
-  return rebinned;
+  if (var.unit() == units::counts ||
+      var.unit() == units::counts * units::counts) {
+    auto dims = var.dimensions();
+    dims.resize(dim, newCoord.dimensions()[dim] - 1);
+    Variable rebinned(var, dims);
+    require<FloatingPointVariableConcept>(rebinned.data())
+        .rebin(var.data(), dim, oldCoord.data(), newCoord.data());
+    return rebinned;
+  } else {
+    // TODO This will currently fail if the data is a multi-dimensional density.
```

Comment:
I'm still not convinced by `rebin` needing to handle both cases of `counts` and `counts/us`. In my opinion, one should only be able to apply `rebin` to counts data and not a distribution. If one needs to `rebin` a distribution, I would say that they first apply the `rebin` and then perform the normalization from bin widths.
I can't really think of a situation where it would not be possible to perform the `rebin` before the normalization.

If `rebin` does one thing and one thing only, then you know exactly what is going into the function and what comes out. This having to check for whether we have counts or distribution might trickle down to other functions and we may find ourselves having to do a lot of units checking in other functions/algorithms.

I would be in favor for having a simple `rebin` function and then maybe define another function `rebin_distribution` or `resample` or something. That could of course internally call `rebin`, but having explicit names that make it clear what each operation does in a workflow would help clarify things.

Of course this comment does not affect the current PR.

Reply:
I think I am at least partially with you:

There are actually three cases, `counts`, `count/us`, and everything else. I agree that it should not be possible to apply `rebin` to the latter (you would use `resample` instead).

For the `counts/us` case, I am not so sure. It is essentially still binned counts, just a different representation. So conceptually it still makes sense to `rebin` it. If we were to not support it, we might consider going one step further and not support `counts/us` at all (rather, it would just be applied on-the-fly, e.g., when plotting), but at this point I am not sure that is a step we'd want to take.